homeassistant/components/fritz/config_flow.py

```diff
@@ -0,0 +1,299 @@
+    ATTR_UPNP_FRIENDLY_NAME,
+    ATTR_UPNP_UDN,
+)
+from homeassistant.config_entries import ConfigFlow
+from homeassistant.const import CONF_HOST, CONF_PASSWORD, CONF_PORT, CONF_USERNAME
+
+from .common import CONFIG_SCHEMA, FritzBoxTools
+from .const import DEFAULT_HOST, DEFAULT_PORT, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@config_entries.HANDLERS.register(DOMAIN)
+class FritzBoxToolsFlowHandler(ConfigFlow):
+    """Handle a FRITZ!Box Tools config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+
+    def __init__(self):
+        """Initialize FRITZ!Box Tools flow."""
+        self._host = None
+        self._entry = None
+        self._name = None
+        self._password = None
+        self._port = None
+        self._username = None
+        self.import_schema = None
+        self.fritz_tools = None
+
+    async def async_step_ssdp(self, discovery_info):
+        """Handle a flow initialized by discovery."""
+        ssdp_location = urlparse(discovery_info[ATTR_SSDP_LOCATION])
+        self._host = ssdp_location.hostname
+        self._port = ssdp_location.port
+        self._name = discovery_info.get(ATTR_UPNP_FRIENDLY_NAME)
+        self.context[CONF_HOST] = self._host
+
+        uuid = discovery_info.get(ATTR_UPNP_UDN)
+        if uuid:
+            if uuid.startswith("uuid:"):
+                uuid = uuid[5:]
+            await self.async_set_unique_id(uuid)
+            self._abort_if_unique_id_configured({CONF_HOST: self._host})
+
+        for progress in self._async_in_progress():
+            if progress.get("context", {}).get(CONF_HOST) == self._host:
+                return self.async_abort(reason="already_in_progress")
+
+        for entry in self.hass.config_entries.async_entries(DOMAIN):
+            if entry.data[CONF_HOST] == self._host:
+                if uuid and not entry.unique_id:
+                    self.hass.config_entries.async_update_entry(entry, unique_id=uuid)
+                return self.async_abort(reason="already_configured")
+
+        self.context["title_placeholders"] = {
+            "name": self._name.replace("FRITZ!Box ", "")
+        }
+        return await self.async_step_confirm()
+
+    async def async_step_confirm(self, user_input=None):
+        """Handle user-confirmation of discovered node."""
+
+        if user_input is None:
+            return self._show_setup_form_confirm()
+
+        errors = {}
+
+        host = self._host
+        port = self._port
+        username = user_input.get(CONF_USERNAME)
+        password = user_input.get(CONF_PASSWORD)
+
+        self.fritz_tools = await self.hass.async_add_executor_job(
+            lambda: FritzBoxTools(
+                hass=self.hass,
+                host=host,
+                port=port,
+                username=username,
+                password=password,
+            )
+        )
+        success, error = await self.hass.async_add_executor_job(self.fritz_tools.is_ok)
```

Comment:
This could be its own function since its done elsewhere as well

Reply:
Done